File: Strings/guess-the-word/guess-the-word.go
// https://leetcode.com/problems/guess-the-word/
package main

type Master struct {
}

func (this *Master) Guess(word string) int {
	return 0
}

func findSecretWord(words []string, master *Master) {
	charCounts := make([]int, 26)
	for _, word := range words {
		for _, char := range word {
			charCounts[char-'a']++
		}
	}

	correctChars := 0

	for correctChars < 6 && len(words) > 0 {
		bestWord := findBestCandidate(words, charCounts)
		correctChars = master.Guess(bestWord)
		words = filterWords(words, bestWord, correctChars, charCounts)
	}
}

func findBestCandidate(words []string, charCounts []int) string {
	maxSum := 0
	bestWord := ""
	for _, word := range words {
		sum := 0
		for _, char := range word {
			sum += charCounts[char-'a']
		}
		if sum > maxSum {
			maxSum = sum
			bestWord = word
		}
	}
	return bestWord
}

func filterWords(words []string, candidate string, candidateCorrectChars int, charCounts []int) []string {
	filteredWords := make([]string, 0)
	for _, word := range words {
		matches := 0
		for i := 0; i < len(candidate); i++ {
            if candidate[i] == word[i] {
                matches++
            }
        }
		if matches == candidateCorrectChars {
			filteredWords = append(filteredWords, word)
		} else {
			for _, char := range candidate {
				charCounts[char-'a']--
			}
		}
	}
	return filteredWords
}

File: Strings/guess-the-word/guess-the-word.rs
// https://leetcode.com/problems/guess-the-word/

struct Master {}

impl Master {
    fn guess(&self, word: String) -> i32 {
        todo!()
    }
}

struct Solution;

impl Solution {
    pub fn find_secret_word(words: Vec<String>, master: &Master) {
        let mut words = words;

        let mut char_counts: [i32; 26] = Self::build_char_counts(&words);

        while let Some(candidate) = Self::find_best_candidate(&words, &char_counts) {
            let correct_chars = master.guess(candidate.clone()) as usize;

            if correct_chars == 6 {
                break;
            }

            words = Self::filter_words(&words, candidate, correct_chars, &mut char_counts)
                .into_iter()
                .cloned()
                .collect();
        }
    }

    fn build_char_counts(words: &[String]) -> [i32; 26] {
        let mut counts = [0; 26];
        for word in words {
            for c in word.bytes() {
                counts[(c - b'a') as usize] += 1;
            }
        }
        counts
    }

    fn filter_words<'a>(
        words: &'a Vec<String>,
        candidate: &String,
        candidate_correct_chars: usize,
        chars: &mut [i32; 26],
    ) -> Vec<&'a String> {
        words
            .iter()
            .filter(|word| {
                let matches = word
                    .bytes()
                    .zip(candidate.bytes())
                    .filter(|(a, b)| a == b)
                    .count();

                if matches != candidate_correct_chars {
                    for c in candidate.bytes() {
                        chars[(c - b'a') as usize] -= 1;
                    }
                }
                matches == candidate_correct_chars
            })
            .collect()
    }

    fn find_best_candidate<'a>(words: &'a [String], char_counts: &[i32; 26]) -> Option<&'a String> {
        words.iter().max_by_key(|word| {
            word.bytes()
                .map(|c| char_counts[(c - b'a') as usize])
                .sum::<i32>()
        })
    }
}

fn main() {
    let words = vec!["acckzz", "ccbazz", "eiowzz", "abcczz"];
    let master = Master {};
    Solution::find_secret_word(words.into_iter().map(|s| s.to_string()).collect(), &master);
}
